crabs/detection_tracking/train_model.py

```diff
@@ -116,7 +118,67 @@ def setup_trainer(self):
             limit_train_batches=self.limit_train_batches,
         )
 
-    def train_model(self):
+    def optuna_maximise_val_precision_recall(
+        self, trial: optuna.Trial
+    ) -> float:
+        """Objective function for Optuna.
+
+        When used with Optuna, it wil maximise precision and recall on the
+        validation set.
+
+        Parameters
+        ----------
+        trial : optuna.Trial
+            The trial to optimise.
+
+        Returns
+        -------
+        float
+            The value to maximise.
+        """
+        # Sample hyperparameters from the search space for this trial
+        # learning rate (a float)
+        if "learning_rate" in self.config["optuna"]:
+            if len(self.config["optuna"]["learning_rate"]) != 2:
+                raise ValueError(
+                    "bounds for `learning_rate` not defined in optuna config"
+                )
+            else:
+                self.config["learning_rate"] = trial.suggest_float(
+                    "learning_rate",
+                    float(self.config["optuna"]["learning_rate"][0]),
+                    float(self.config["optuna"]["learning_rate"][1]),
+                )
+
+        # num_epochs (an integer)
+        if "num_epochs" in self.config["optuna"]:
+            if len(self.config["optuna"]["num_epochs"]) != 2:
+                raise ValueError(
+                    "bounds for `num_epochs` not defined in optuna config"
+                )
+            else:
+                self.config["num_epochs"] = trial.suggest_int(
+                    "num_epochs",
+                    self.config["optuna"]["num_epochs"][0],
+                    self.config["optuna"]["num_epochs"][1],
+                )
+
+        # Run training
+        trainer = self.core_training()
+
+        # Return metric to maximise
+        val_precision = trainer.callback_metrics["val_precision"].item()
+        val_recall = trainer.callback_metrics["val_recall"].item()
+        return (val_precision + val_recall) / 2
+
+    def core_training(self) -> lightning.Trainer:
```

Comment:
suggestions for a better name for this function are welcome 😬

Reply:
The reason the optuna function is not in train_model is to avoid this file getting too long. I feel like main file, should be short

Reply:
yeah agreed, I only kept the bits that are very specific to this training and kept the general optuna parts in the optuna module, but happy to discuss.

Reply:
What is wrong with the one before? I think having optuna objective in `train_model.py` will just make it messy, as optuna option is an enhancement. It is not something people use frequently (as it takes at least a week to get decent study done).

Reply:
I agree optuna is an enhancement and people won't use it frequently. In this PR I was trying to reduce [this code duplication](https://github.com/SainsburyWellcomeCentre/crabs-exploration/blob/d3f17e41f0326d81910a287c521936d7bc27116e/crabs/detection_tracking/optuna_fn.py#L83-L106).

My thinking was that it is more confusing to have repeated parts of the training code in different modules.